frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -303,17 +326,43 @@ export default ({
         }
       })
     },
-    setInitMessages () {
-      if (this.isJoinedChatRoom(this.currentChatRoomId)) {
-        this.messages = cloneDeep(this.chatRoomLatestMessages)
+    getSimulatedState (initialize = true) {
+      return {
+        settings: cloneDeep(this.chatRoomSettings),
+        attributes: cloneDeep(this.chatRoomAttributes),
+        users: cloneDeep(this.chatRoomUsers),
+        messages: initialize ? [] : this.messages,
+        simulation: true
+      }
+    },
+    async getLatestEvents (refresh = false) {
+      const limit = this.chatRoomSettings?.actionsPerPage || CHATROOM_ACTIONS_PER_PAGE
+      const before = refresh || !this.latestEvents.length
+        ? ''
+        : GIMessage.deserialize(this.latestEvents[0]).hash()
+
+      const newEvents = await sbp('chelonia/contractEventsBefore', this.currentChatRoomId, before, limit)
+
+      if (refresh) {
+        this.latestEvents = newEvents
       } else {
-        this.messages = []
-        sbp('okTurtles.events/once', `${CHATROOM_STATE_LOADED}-${this.currentChatRoomId}`, (state) => {
-          this.messages = getLatestMessages({
-            count: this.chatRoomSettings.messagesPerPage, // TODO: this.chatRoomSettings could be {}
-            messages: state.messages
-          })
-        })
+        this.latestEvents.splice(0, 0, ...newEvents)
```

Comment:
A bit cleaner I think: `this.latestEvents.unshift(...newEvents)`

Reply:
Perfect. That's correct.